### DISCUSSION TITLE
Shuffled Tx Order – Dispenser Solution?

### DISCUSSION THREAD

**Discussion by Jpja** [CounterpartyXCP/CIPs#122] [CIPs]
I throw this out as an alternative solution since [Dispenser Reservation](https://github.com/CounterpartyXCP/cips/discussions/121) is unlikely to get implemented anytime soon.

Counterparty logic is first included in a block, first served. Default Bitcoin miner policy is the higher the fee ratio (sat/byte), the earlier included in a block. 

Scammers take advantage of this by frontrunning incoming dispenses. They simply cancel the dispenser (or send a competing dispense) with a slightly higher fee, so the victim loses his BTC without receiving the expected token.

Legitimate buyers may also end up in a "RBF war" where two buyers up their fees to win the dispenser (I have not seen such behavior yet, but expect it to happen).

**Solution:** Counterparty shuffles all dispenses within a block. If there are two incoming dispenses but only one lot available, each dispense has an equal 50% chance of getting filled. Three incoming dispenses, and the probability is 1/3, etc. A second rule is that cancellations get the lowest priority, so the seller can't just close the dispenser in front of the buyer.

When Counterparty processes a block, it will (instead of dispensing immediately) assign a potential dispense a score. The score comes from a random function with the block hash as the seed. This makes the score unpredictable before the block is mined, yet using a seed means all nodes will come up with the exact same score. A cancel is given a score of 0.

After all other transactions in the block is parsed, Counterparty starts going through the list of potential dispenses. It works its way from the highest score down.

**Concerns:** 
- This is deep level change as it messes with tx processing.
- An attacker can send 2 or 3 or .. N transactions to try to front run. The probability of success is to N/(1+N). The cost is tx fees and liquidity. Think of an attack where a ₿0.1 token is listed. Attacker sees incoming dispenses and immediately sends 9 competing dispenses. The cost is 9 fees (~$5) an ₿0.9 of liquidity to pull this off. The probability of successful attack is 90%.

To be clear, this will only make the green and orange areas of the graph slightly larger. I don't think it's worth messing with the core code for this.

https://user-images.githubusercontent.com/8447505/272190358-8f48afbb-2f66-4b12-b5f1-2d5cb6ca63ee.png

**Comment by jdogresorg** [comment]
This is an interesting proposal, and perhaps a feature we could/should add to dispensers in the future, but I dont think this should be the default behavior.

Since its inception, CP has always processed transactions in the order that they were put in the block (usually highest miners fee first), so I'm not sure that we should depart from that model as the standard.

IMO If txs are competing for a single dispense, the one that should get the dispenses is the first one included in the block, if we depart from that model, we should make it very obvious, and make it an optional "feature" of a dispenser.

I like the idea of "random" dispenses and dispensing of "random" cards / packs, etc.... Def want to see some "random" features in dispensers in the future :)
